### PR TITLE
[KernelGen][Nvidia] Add trapz operator with Triton kernel

### PR DESCRIPTION
**Pull request by NineAnnAnn**
## Summary
Adds a Triton kernel for `trapz`. Computes trapezoidal rule integration along a dimension using constant spacing dx.

## Testing
- Parametrized tests over shapes (1D, 2D, 3D), dtypes (float32, float16, bfloat16), dx values, and dimensions
- Validated against reference on device via `to_reference(inp, True)`
- Edge cases: single element, two elements, large reduction dimensions (1024, 2048)

## Performance
Test command: `pytest benchmark/test_trapz.py --level core` (NVIDIA H20)

### trapz

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [64, 64] | 0.015 | 0.008 | 1.984 |
| float16 | [4096, 4096] | 0.037 | 0.052 | 0.719 |
| float16 | [64, 512, 512] | 0.046 | 0.045 | 1.017 |
| float16 | [1024, 1024, 1024] | 3.566 | 2.438 | 1.463 |
| float16 | [10000, 1] | 0.017 | 0.005 | 3.547 |
| float16 | [10000, 256] | 0.025 | 0.014 | 1.740 |
| float16 | [10000, 65536] | 0.522 | 4.048 | 0.129 |
| float16 | [100, 1, 100] | 0.017 | 0.009 | 1.827 |
| float16 | [100, 256, 100] | 0.032 | 0.021 | 1.542 |
| float16 | [100, 65536, 100] | 4.129 | 5.209 | 0.793 |
| float32 | [64, 64] | 0.015 | 0.006 | 2.684 |
| float32 | [4096, 4096] | 0.045 | 0.058 | 0.765 |
| float32 | [64, 512, 512] | 0.054 | 0.045 | 1.201 |
| float32 | [1024, 1024, 1024] | 2.138 | 2.599 | 0.822 |
| float32 | [10000, 1] | 0.016 | 0.005 | 3.397 |
| float32 | [10000, 256] | 0.024 | 0.013 | 1.876 |
| float32 | [10000, 65536] | 0.706 | 2.068 | 0.341 |
| float32 | [100, 1, 100] | 0.017 | 0.007 | 2.432 |
| float32 | [100, 256, 100] | 0.033 | 0.018 | 1.821 |
| float32 | [100, 65536, 100] | 8.392 | 3.031 | 2.768 |
| bfloat16 | [64, 64] | 0.015 | 0.008 | 1.909 |
| bfloat16 | [4096, 4096] | 0.037 | 0.051 | 0.734 |
| bfloat16 | [64, 512, 512] | 0.046 | 0.043 | 1.075 |
| bfloat16 | [1024, 1024, 1024] | 1.509 | 2.479 | 0.609 |
| bfloat16 | [10000, 1] | 0.018 | 0.005 | 3.815 |
| bfloat16 | [10000, 256] | 0.025 | 0.014 | 1.734 |
| bfloat16 | [10000, 65536] | 0.523 | 4.046 | 0.129 |
| bfloat16 | [100, 1, 100] | 0.017 | 0.010 | 1.618 |
| bfloat16 | [100, 256, 100] | 0.032 | 0.114 | 0.285 |
| bfloat16 | [100, 65536, 100] | 8.280 | 3.159 | 2.621 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| trapz | **1.580** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (30 cases) | 1.580 | Primary |
| Tianshu | N/A | — | Not tested |
| Muxi | N/A | — | Not tested |
| Ascend | N/A | — | Not tested |
| Hygon | N/A | — | Not tested |

## Files Changed
- `src/flag_gems/ops/trapz.py`: Triton kernel implementation
- `tests/test_trapz.py`: Accuracy test
- `benchmark/test_trapz.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.1)

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -13889,6 +13889,24 @@ ops:
...
+  for:
+  - trapezoid.dx
+  - trapz.dx
+  labels:
+  - aten
+  - KernelGen
+  kind:
+  - Math
+  stages:
+  - alpha: '5.1'
```
Comment (Caeruleann) [inline conf/operators.yaml:13909]:
```suggestion
  - alpha: '5.4'
```

File: src/flag_gems/ops/trapz.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    """Compute trapezoid integration with constant spacing dx."""
+    logger.debug("GEMS TRAPEZOID_DX")
+
+    # Handle empty tensor
+    if y.numel() == 0:
+        shape = list(y.shape)
+        dim = dim % y.ndim if y.ndim > 0 else 0
+        if y.ndim > 0:
+            shape[dim] = 0
+        return torch.empty(shape, dtype=y.dtype, device=y.device)
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:148]: Please preserve the reduction shape for empty inputs. For example, (2, 0, 3) with dim=1 should return (2, 3), while the current early return keeps the reduced dimension and returns uninitialized data.

File: src/flag_gems/ops/trapz.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    shape = list(y.shape)
+    dim = dim % y.ndim
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:151]: Please validate dim before normalization. Using modulo silently accepts out-of-range dimensions such as dim=2 for a 2-D input instead of matching PyTorch’s error behavior.

File: src/flag_gems/ops/trapz.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    # If dimension size is 0 or 1, return zeros with reduced dimension
+    if N <= 1:
+        out_shape = shape[:dim] + shape[dim + 1 :]
+        return torch.zeros(out_shape, dtype=y.dtype, device=y.device)
+
+    # Compress dimensions for efficient processing
+    y_compressed = dim_compress(y, dim)
+    M = y_compressed.numel() // N
+
+    # Handle x tensor - check if it's 1D or same shape as y
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:213]: trapezoid_x can read out of bounds for native-supported broadcast inputs such as y.shape=(2, 3) and x.shape=(1, 3), because the kernel iterates over M rows while x only contains one row.

File: src/flag_gems/ops/trapz.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    # Reshape back to original dimensions and convert to original dtype
+    out_shape_final = shape[:dim] + shape[dim + 1 :]
+    result = output.reshape(out_shape_final)
+    if compute_dtype != y.dtype:
+        result = result.to(y.dtype)
+    return result
+
+
+def trapezoid_x(y, x, dim=-1):
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:188]: This function is not registered and currently has no test coverage, so please either remove it or add full broadcasting support with ATen parity tests.

File: src/flag_gems/ops/trapz.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+        # Sum across N dimension
+        acc += tl.sum(trapezoid_sum, axis=1)
+
+    # Store result (keep in float32)
+    out_offsets = m_offsets * M_STRIDE_OUT
+    tl.store(out_ptr + out_offsets, acc, mask=mask_m)
+
+
+def trapz(y, dx=1.0, dim=-1):
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:138]: Please preserve PyTorch dtype and autograd contracts: integer inputs require the correct promotion, FP64 must not be accumulated in FP32, complex inputs need support or a safe fallback, bool must be rejected, and differentiable inputs must retain backward support via a custom autograd implementation or differentiable FlagGems operators.

File: tests/test_trapz.py
Comment (Caeruleann) [inline tests/test_trapz.py:1]: Please extend the ATen parity tests to cover empty reduced/free dimensions, scalar and out-of-range dimensions, non-contiguous inputs, integer/FP64/complex/bool dtypes, native-supported broadcast x forms, and forward/backward gradient correctness.

File: src/flag_gems/ops/trapz.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    # Triton kernels are real-valued only; defer complex inputs to PyTorch's
+    # native implementation (on CPU, since the CUDA dispatch key is overridden
+    # by FlagGems and a direct torch.trapezoid call would re-enter this op).
+    # complex32 lacks native reduction kernels on CPU, so upcast it first.
+    if y.is_complex():
+        comp_dtype = (
+            torch.complex128 if y.dtype == torch.complex128 else torch.complex64
+        )
+        out = torch.trapezoid(y.cpu().to(comp_dtype), dx=dx, dim=dim)
+        return out.to(y.dtype).to(y.device)
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:189]: Please reject complex inputs with an explicit error if they are unsupported, rather than falling back to the native ATen implementation.

File: src/flag_gems/ops/trapz.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    configs=runtime.get_tuned_config("trapezoid_kernel"),
+    key=["M", "N"],
+)
+@triton.jit
+def trapezoid_kernel_dx(
+    y_ptr,
+    out_ptr,
+    M,
+    N,
+    dx,
```
Comment (Caeruleann) [inline src/flag_gems/ops/trapz.py:41]: Triton passes the Python-float `dx` as an fp32 scalar, so fp64 inputs lose precision at the final `acc * dx` (measured 1.49e-8 rel err with dx=0.1 — exactly fp32 rounding of dx; dx=0.5 is lossless). Please annotate `dx: tl.float64` (or pass a 1-elem fp64 tensor) and add an fp64 test with a non-exactly-representable dx such as 0.1.

File: conf/operators.yaml [hunk 9]
```diff
@@ -16591,6 +16532,21 @@ ops:
...
+  for:
+    - trapezoid.dx
+    - trapz.dx
+  labels:
+    - aten
+    - KernelGen
+  kind:
+    - Math
+  stages:
+    - alpha: '5.4'
```
Comment (Caeruleann) [inline conf/operators.yaml]:
```suggestion
    - alpha: '5.5'
```

**Review by Caeruleann — APPROVED** [review]
